File: src/leetcode/java/AddBinary_67.java
package leetcode.java;

public class AddBinary_67 {
  public String addBinary(String a, String b) {
    char[] left = a.length() >= b.length() ? a.toCharArray() : b.toCharArray();
    char[] right = a.length() < b.length() ? a.toCharArray() : b.toCharArray();
    boolean carry = false;
    int i = left.length - 1, j = right.length - 1;
    StringBuilder res = new StringBuilder();
    for (; j >= 0; j--, i--) carry = binary(left[i], right[j], carry, res);
    for (; i >= 0; i--) carry = binary(left[i], '0', carry, res);
    if (carry) res.append(1);
    return res.reverse().toString();
  }

  private static boolean binary(char c, char d, boolean carry, StringBuilder res) {
    int s = (c - '0') + (d - '0') + (carry ? 1 : 0);
    res.append(s % 2);
    return s >= 2;
  }
}

File: src/leetcode/python/AddBinary_67.py
class AddBinary_67(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        if len(a) < len(b):
            a, b = b, a
        res = []

        def binary(l, r, c):
            s = int(l) + int(r) + (1 if c else 0)
            res.append(str(s % 2))
            return s >= 2

        bl, br = len(a) - 1, len(b) - 1
        c = False
        while br >= 0:
            c = binary(a[bl], b[br], c)
            bl, br = bl - 1, br - 1
        while bl >= 0:
            c = binary(a[bl], '0', c)
            bl -= 1
        if c:
            res.append("1")
        return ''.join(res[::-1])
